### DISCUSSION TITLE
Teleproto using chimney macros

### DISCUSSION THREAD

**Discussion by urmaul** [scalalandio/chimney#573] [General]
This is a placeholder for discussions about migrating teleproto to use chimney macros inside.

WIP draft implementation for this can be found here: https://github.com/moia-oss/teleproto/pull/338

**Comment by MateuszKubuszok** [comment]
Soo, this PR still relies on using `Transformer`s (and maybe `PartialTransformer`s) as the basis.

What I had in mind is more like this:

```scala
package io.scalaland.chimney.internal.compiletime

import io.scalaland.chimney.internal.compiletime.{datatypes, ChimneyDefinitionsPlatform}

trait ChimneyAsLibrary
    extends Derivation
    with ChimneyDefinitionsPlatform
    with datatypes.IterableOrArraysPlatform
    with datatypes.ProductTypesPlatform
    with datatypes.SealedHierarchiesPlatform
    with datatypes.ValueClassesPlatform
    with rules.TransformImplicitRuleModule
    with rules.TransformSubtypesRuleModule
    with rules.TransformToSingletonRuleModule
    with rules.TransformOptionToOptionRuleModule
    with rules.TransformPartialOptionToNonOptionRuleModule
    with rules.TransformToOptionRuleModule
    with rules.TransformValueClassToValueClassRuleModule
    with rules.TransformValueClassToTypeRuleModule
    with rules.TransformTypeToValueClassRuleModule
    with rules.TransformEitherToEitherRuleModule
    with rules.TransformMapToMapRuleModule
    with rules.TransformIterableToIterableRuleModule
    with rules.TransformProductToProductRuleModule
    with rules.TransformSealedHierarchyToSealedHierarchyRuleModule
```

```scala
package io.moia.protos.teleproto.internal

import io.scalaland.chimney.internal.compiletime.ChimneyAsLibrary

import scala.reflect.macros.blackbox

class EncoderMacros(val c: blackbox.Context)
    extends io.scalaland.chimney.internal.compiletime.ChimneyAsLibrary
    with ... /* teleprotos own rules e.g. EncodeImplicitRule */ {

  def deriveEncoderWithDefaults[
      From: WeakTypeTag,
      To: WeakTypeTag
  ]: Expr[Encoder[From, To]] = {
    // pseudocode:

    val inputName = freshTermName(...)
    val inputExpr: Expr[From] = c.Expr[From](q"$inputName")

    val cfg = TransformerConfiguration() // customize, read config with DSL etc
    val context = TransformationContext.ForTotal(inputExpr, cfg)
    
    deriveFinalTransformationResultExpr(context).toEither.fold(
      derivationErrors => /* report somehow */,
      toExpr =>
        c.Expr[Encoder[From, To](
          q"""
          new Encoder[${Type[From]}, ${Type[To]}] {
            def encode($inputName: ${Type[From]}): ${Type[To]} = $toExpr
          }
          """
        )
    )
  }
  
  // chain of responsibility of sort: attempt first rule, it it yields another, then another, etc
  final override protected val rulesAvailableForPlatform: List[Rule] = List(
    // TransformImplicitRule, // looks for Transformer/PartialTransformer, instead provide your own:
    EncodeImplicitRule, // <- this should look for implicit Encoder
    TransformSubtypesRule,
    TransformToSingletonRule,
    TransformOptionToOptionRule,
    TransformPartialOptionToNonOptionRule,
    TransformToOptionRule,
    TransformValueClassToValueClassRule,
    TransformValueClassToTypeRule,
    TransformTypeToValueClassRule,
    TransformEitherToEitherRule,
    TransformMapToMapRule,
    TransformIterableToIterableRule,
    TransformProductToProductRule,
    TransformSealedHierarchyToSealedHierarchyRule
  )
}
```

Something like this would be able to:

 - remove dependency on `Transformer` and `PartialTransformer` type classes
 - allow injecting custom rules
 - remove the need of vampyre macros or similar
 - customizing the error message

**Reply by MateuszKubuszok** [reply]
It was just a quick pseudocode but I wanted to give a general idea.

**Reply by urmaul** [reply]
Thank you, I'll look into it. My idea was to copypaste the outermost macro and then iteratively dig deeper while keeping it compiling and working all the time. That's probably more work but less big changes and opportunities to get lost.

**Reply by MateuszKubuszok** [reply]
Basically Chimney at this point has a customizable derivation engine of sort:

 - config, input, overrides are passed around with `TransformerContext`
 - derivation `Rule` make a chain of responsibility where the first match creates result, rule can also fail or yield
 - `DerivationResult` is like Either (of non empty list of failures) + Writer (logs)
 - whole thing take `src: Expr[From]` and returns `Expr[To]` or `Expr[partial.Result[To]]` wrapping with a type class is done in the entrypoint to the macro

So I would see teleproto creating the context on its own, reusing some rules and creating a few on its own (e.g. summoning implicit Encoder/Decoder) and having its own entrypoint calling the engine and wrapping its output.

**Comment by MateuszKubuszok** [comment]
I've updated the draft a bit:

`src/main/scala/io/scalaland/chimney/internal/compiletime/ExportChimneyDerivationEngineForTeleproto.scala` - part of the macro that can be shared between Scala 2 and Scala 3, exposing Chimney's parts that are relevant to derivation (with a name that would prevents conflicts) 

```scala
package io.scalaland.chimney.internal.compiletime

trait ExportChimneyDerivationEngineForTeleproto
    extends derivation.transformer.Derivation
    with derivation.transformer.Configurations
    with derivation.transformer.Contexts
    with derivation.transformer.ImplicitSummoning
    with derivation.transformer.ResultOps
    with datatypes.IterableOrArrays
    with datatypes.ProductTypes
    with datatypes.SealedHierarchies
    with datatypes.SingletonTypes
    with datatypes.ValueClasses
    with derivation.transformer.integrations.OptionalValues
    with derivation.transformer.integrations.PartiallyBuildIterables
    with derivation.transformer.integrations.TotallyBuildIterables
    with derivation.transformer.integrations.TotallyOrPartiallyBuildIterables
    with derivation.transformer.rules.TransformationRules
```

`src/main/scala-2/io/scalaland/chimney/internal/compiletime/ExportChimneyDerivationEngineForTeleproto.scala` - Scala 2-specific implementations of abstract interfaces, using Scala 2 macros:

```scala
package io.scalaland.chimney.internal.compiletime

trait ExportChimneyDerivationEngineForTeleprotoPlatform
  extends ExportChimneyDerivationEngineForTeleproto
    with ChimneyDefinitionsPlatform
    with datatypes.IterableOrArraysPlatform
    with datatypes.ProductTypesPlatform
    with datatypes.SealedHierarchiesPlatform
    with datatypes.ValueClassesPlatform
    with derivation.transformer.rules.TransformImplicitRuleModule
    with derivation.transformer.rules.TransformSubtypesRuleModule
    with derivation.transformer.rules.TransformToSingletonRuleModule
    with derivation.transformer.rules.TransformOptionToOptionRuleModule
    with derivation.transformer.rules.TransformPartialOptionToNonOptionRuleModule
    with derivation.transformer.rules.TransformToOptionRuleModule
    with derivation.transformer.rules.TransformValueClassToValueClassRuleModule
    with derivation.transformer.rules.TransformValueClassToTypeRuleModule
    with derivation.transformer.rules.TransformTypeToValueClassRuleModule
    with derivation.transformer.rules.TransformEitherToEitherRuleModule
    with derivation.transformer.rules.TransformMapToMapRuleModule
    with derivation.transformer.rules.TransformIterableToIterableRuleModule
    with derivation.transformer.rules.TransformProductToProductRuleModule
    with derivation.transformer.rules.TransformSealedHierarchyToSealedHierarchyRuleModule
```

`src/main/scala-3/io/scalaland/chimney/internal/compiletime/ExportChimneyDerivationEngineForTeleproto.scala` - Scala 2-specific implementations of abstract interfaces, using Scala 3 macros:

```scala
package io.scalaland.chimney.internal.compiletime

// using q: Quotes, require us to make it abstract class on Scala 3 to make it a stable identifier
abstract class ExportChimneyDerivationEngineForTeleprotoPlatform(q: scala.quoted.Quote)
  extends ChimneyDefinitionsPlatform(q)
    with ExportChimneyDerivationEngineForTeleproto
    with datatypes.IterableOrArraysPlatform
    with datatypes.ProductTypesPlatform
    with datatypes.SealedHierarchiesPlatform
    with datatypes.ValueClassesPlatform
    with derivation.transformer.rules.TransformImplicitRuleModule
    with derivation.transformer.rules.TransformSubtypesRuleModule
    with derivation.transformer.rules.TransformToSingletonRuleModule
    with derivation.transformer.rules.TransformOptionToOptionRuleModule
    with derivation.transformer.rules.TransformPartialOptionToNonOptionRuleModule
    with derivation.transformer.rules.TransformToOptionRuleModule
    with derivation.transformer.rules.TransformValueClassToValueClassRuleModule
    with derivation.transformer.rules.TransformValueClassToTypeRuleModule
    with derivation.transformer.rules.TransformTypeToValueClassRuleModule
    with derivation.transformer.rules.TransformEitherToEitherRuleModule
    with derivation.transformer.rules.TransformMapToMapRuleModule
    with derivation.transformer.rules.TransformIterableToIterableRuleModule
    with derivation.transformer.rules.TransformProductToProductRuleModule
    with derivation.transformer.rules.TransformSealedHierarchyToSealedHierarchyRuleModule
```

Then in teleproto:

`src/main/scala/io/moia/protos/teleproto/internal/TeleprotoDefinitions.scala` - use it as `this: TeleprotoDefinitions => ` when defining your own cross-compilable mixins:

```scala
package io.moia.protos.teleproto.internal

trait TeleprotoDefinitions
  extends io.scalaland.chimney.internal.compiletime.ExportChimneyDerivationEngineForTeleproto
  // ...
  {

  // definitions of teleproto types e.g. Encoder, Decoder

  // interfaces for e.g. summoning implicit Encoder, Decoder

  // code which would parse teleproto configs and turn them into TransformationContext for Chimney
}
```

`src/main/scala-2/io/moia/protos/teleproto/internal/TeleprotoDefinitionsPlatform.scala`:

```scala
package io.moia.protos.teleproto.internal

trait TeleprotoDefinitionsPlatform
  extends TeleprotoDefinitions
  with io.scalaland.chimney.internal.compiletime.ExportChimneyDerivationEngineForTeleprotoPlatform {

  // implementations of teleproto types e.g. Encoder, Decoder

  // implementations for e.g. summoning implicit Encoder, Decoder
}
```

```scala
package io.moia.protos.teleproto.internal

import scala.reflect.macros.blackbox

class TeleprotoMacros(val c: blackbox.Context)
    extends TeleprotoDefinitionsPlatform
    with ... /* teleprotos own rules e.g. EncodeImplicitRule */ {

  def deriveEncoderWithDefaults[
      From: WeakTypeTag,
      To: WeakTypeTag
  ]: Expr[Encoder[From, To]] = {
    // pseudocode:

    val inputName = freshTermName(...)
    val inputExpr: Expr[From] = c.Expr[From](q"$inputName")

    val cfg = TransformerConfiguration() // customize, read config with DSL etc
    val context = TransformationContext.ForTotal(inputExpr, cfg)
    
    deriveFinalTransformationResultExpr(context).toEither.fold(
      derivationErrors => /* report somehow */,
      toExpr =>
        c.Expr[Encoder[From, To](
          q"""
          new Encoder[${Type[From]}, ${Type[To]}] {
            def encode($inputName: ${Type[From]}): ${Type[To]} = $toExpr
          }
          """
        )
    )
  }

  def deriveEncoderWithDefaults[
      From: WeakTypeTag,
      To: WeakTypeTag
  ]: Expr[Encoder[From, To]] = {
    // pseudocode:

    val inputName = freshTermName(...)
    val inputExpr: Expr[From] = c.Expr[From](q"$inputName")

    val cfg = TransformerConfiguration() // customize, read config with DSL etc
    val context = TransformationContext.ForPartial(inputExpr, cfg)
    
    deriveFinalTransformationResultExpr(context).toEither.fold(
      derivationErrors => /* report somehow */,
      toExpr =>
        c.Expr[Encoder[From, To](
          q"""
          new Encoder[${Type[From]}, ${Type[To]}] {
            def encode($inputName: ${Type[From]}): ${Type[To]} = ${convertPartialToPbResult(toExpr)}
          }
          """
        )
    )
  }
  
  // chain of responsibility of sort: attempt first rule, it it yields another, then another, etc
  final override protected val rulesAvailableForPlatform: List[Rule] = List(
    // TransformImplicitRule, // looks for Transformer/PartialTransformer, instead provide your own:
    EncodeImplicitRule, // <- this should look for implicit Encoder
    TransformSubtypesRule,
    TransformToSingletonRule,
    TransformOptionToOptionRule,
    TransformPartialOptionToNonOptionRule,
    TransformToOptionRule,
    TransformValueClassToValueClassRule,
    TransformValueClassToTypeRule,
    TransformTypeToValueClassRule,
    TransformEitherToEitherRule,
    TransformMapToMapRule,
    TransformIterableToIterableRule,
    TransformProductToProductRule,
    TransformSealedHierarchyToSealedHierarchyRule
  )
}
```

To understand this architecture:

 * basically whole derivation logic is macro agnostic
 * one is defining interfaces (including Expr, Type) and methods operating on them and whole engine only uses these types and methods
 * we add the implementation at the end of the world with mix-ins (so the whole macro is a giant cake pattern, kinda unavoidable with path-dependent types)
 * the limitation is that `deriveFinalTransformationResultExpr(context)` can only derive Expr[To] (total) or `Expr[partial.Result[To]]`. The former works for `Encoder`s, but `Decoder`s would have to convert `PbResult` into `partial.Result` within the macro (e.g. converting result of a decoding with implicit `Decoder`), and from `partial.Result` to `PbResult` at the exit from the macro
 * it's a bit more complicated than it needs to be since, it's the first time someone reuses the engine, and a lot of interfaces are package-private to not pollute the intellisense. If this approach works out, we could expose these traits in some module I guess?

**Comment by MateuszKubuszok** [comment]
I drafted [this PR](https://github.com/scalalandio/chimney/pull/574) as an example how Chimney can be use in another library with different type classes, and how to make it cross-compilable between Scala 2 and Scala 3.

If this approach makes sense, we could release `chimney-engine` as an additional module intended for other libraries to use, but the way we implement cross-compilation kinda forces the architecture on any adopter.

**Reply by urmaul** [reply]
I shamelessly/mindlessly copypasted the code from your PR and it works great except the one case. I get the endless recursion in `OneOfProtocolBuffersTest` and I'm not sure is it is a bug in my code or an unexpected edge case in chimney macros. Is there anything super-special in `TransformSealedHierarchyToSealedHierarchyRule`?

**Reply by MateuszKubuszok** [reply]
Let me check, what it actually does:

 1. it only attempts to derive if both source and target are: sealed, Scala 3 enum or Java enum (nothing suspicious)
 2. at first it tries to find manual overrides, use them, and remove them from the list of things to handle (nothing suspicious either)
 3. then it attempts matching remaining subtypes by name:
    - 0 matches -> error (nothing suspicious)
    - 2+ matches -> error (same)
    - exactly 1 match - it attempts to convert input subtype into output subtype:
      a. at first it tries to use `FromSubtype <: From`, `ToSubtype <: To` recursive derivation - with all the rules (implicits, subtypes, singletons, anyvals, products, ...), it shouldn't cause infinite recursion, since types are different
      b. then it has a fallback rules for wrappers in PBs (`case Foo.A(a) => /* Bar.A from unwrapped a */`) - I DID wrote some unit tests to make sure it would not interfere with other rules
      c. same for `case a: Foo.A => Bar.A(/* Bar.A's only value from Foo.A */)` - this one too I unit tested to make sure it is not interfering with other rules
 4. then it makes the last resort attempt - `FromSubtype` into `To` using `Implicit` rule, it allows us to provide some `Transforerm[FromSubtype, To]` and let it be detected - it should not interfere with anything UNLESS you are providing some generic autoderivation that would be caught by this
   - I remember that I did have a problem before filtering rules was added (so that we started to consider only `Implicit` rule): `SealedToSealed` matched `From` into `To` -> special case `FromSubtype into To` was used -> then `FromSubtype` got upcasted into `From` -> recursive derivation got to `SealedToSealed` again... But now it should not be an issue, since we separate user-provided transformers and autoderived transformers.

I would take check:

 1. is there a separation between automatic derivation and semiautomatic (engine kind of assumes that users only provide `Transformer`s/`PartialTransformer`s and automatic derivation will NOT be triggered recursively since `Transformers.AutoDerived` and `PartialTransformer.AutoDerived` are supertypes)
 2. if any added rule also stick to this convention - I think I cut some corners when writing the demo PR, so if one were to copy-paste is and just add `given` to the `inline def derived`, the code might trigger autoderivation even though the engine assume it would not, so it might require some fixing:

    ```scala
    package io.scalaland.chimney.example

    // API that users should be using.
    trait MyTypeClass[From, To] extends MyTypeClass.AutoDerived[From, To] {

      def convert(from: From): To
    }
    object MyTypeClass extends MyTypeClassCompanionPlatform {
  
      // Part of the ABI, but not one that users should be aware of,
      // only intended to be used by extension methods.
      trait AutoDerived[From, To] {

        def convert(from: From): To
      }
      object AutoDerived extends MyTypeClassAutoDerivedCompanionPlatform
    }
    ```

    ```scala
    // Scala 3 version, but Scala 2 would work identically

    trait MyTypeClassCompanionPlatform { this: MyTypeClass.type =>

      // NOT given/implicit:
      // should be used for semiautomatic derivation and user-provided types.
      inline def deriving[From, To]: MyTypeClass[From, To] =
        ${ internal.MyTypeClassMacros.derivingImpl[From, To] }
    }


    trait MyTypeClassAutoDerivationCompanionPlatform { this: MyTypeClass.AutoDerived.type =>

      // Indeed given/implicit:
      // should be used by extension methods to look for BOTH user-provided conversion
      // as well as automatically derived instances.
      //
      // It reuses macro, but upcast the results so that derivation would not pick this
      // implicit/given by accident.
      inline given deriving[From, To]: MyTypeClass.AutoDerived[From, To] =
        ${ internal.MyTypeClassMacros.derivingImpl[From, To] }
    }
    ```

**Reply by urmaul** [reply]
> I remember that I did have a problem before filtering rules was added (so that we started to consider only Implicit rule): SealedToSealed matched From into To -> special case FromSubtype into To was used -> then FromSubtype got upcasted into From -> recursive derivation got to SealedToSealed again... But now it should not be an issue, since we separate user-provided transformers and autoderived transformers.

Turned out this was the reason in one case. In the test code we derived the transformer for subtype and added it to implicit scope. Commenting this out fixed the issue.

There is another test where similar thing happens. We have a writer defined like this:

```
implicit val writer: Writer[Meal, food.Meal] = ProtocolBuffers.writer[Meal, food.Meal]
```

It compiles fine and I can even try to print it (prints something like `io.moia.protos.teleproto.ProtocolBuffersRoundTripTest$$Lambda/0x0000007001194018@58695725`) but if I try to use it, I get StackOverflow error.

In this case print statements in my `WriterImplicitRule` show that my derived writer found itself in the implicit scope. And there is `TransformProductToProductRuleModule.scala` deep in the stack trace. Probably the cause is something simple too, I have to debug more.

**Reply by MateuszKubuszok** [reply]
In our implicit rule we created special handling:

 * when we create the context for the outermost derivation we [disable summoning](https://github.com/scalalandio/chimney/blob/a4db6652e0a2639ffec735a88fd412dc8acb4080/chimney/src/main/scala/io/scalaland/chimney/internal/compiletime/derivation/transformer/Configurations.scala#L294) for semiautomatic derivation
 * when we attempt summoning we [check if we should be able to do it](https://github.com/scalalandio/chimney/blob/a4db6652e0a2639ffec735a88fd412dc8acb4080/chimney/src/main/scala/io/scalaland/chimney/internal/compiletime/derivation/transformer/ImplicitSummoning.scala#L11)
 * then top-level derivation moves on to the next rule, and circular dependency on onself is prevented
 * when going into recursive derivation, we're remove that ["do not summon" marker](https://github.com/scalalandio/chimney/blob/a4db6652e0a2639ffec735a88fd412dc8acb4080/chimney/src/main/scala/io/scalaland/chimney/internal/compiletime/derivation/transformer/Configurations.scala#L391)

So the assumption is that:

 * you do not use automatic derivation when there can be a cycle
 * you allow semiautomatic derivation which would prevent resolving implicit to oneself, but allow it in nested contexts

      ```scala
      case class Foo(a: Option[Foo])
      case class Bar(a: Option[Bar])
      implicit def value = new Transformer[Foo, Bar] {
        def transform(src: Foo): Bar = 
          // here we cannot allow value.transform(src)
          new Bar(src.a.map(foo => value.transform(foo)) // here we have to allow value.transform(foo)
      }
      ```

**Reply by urmaul** [reply]
Turned out I see infinite recursion only if we fail to derive a macro. I still have to fix that but this way I can leave it to future me and do somthing else instead 😈

Regarding something else, maybe there's a simple thing I don't see. I'm trying to make a derivation rule for enums that can handle class name as prefix in the variant names. Is there a way to extract the class name from `Type[To]`? I see prettyPrint does something similar but it adds formatting I don't need.

**Reply by MateuszKubuszok** [reply]
There is no such utility right now. Enums in particular are returned as a `ListMap` when using SealedHierarchies utility, and they have their name as a key of the list, but no way of obtaining such name OOTB.

I made a #600 for this. In some other project where I used chimney-macro-commons I temporarily worked around this limitation by:

```scala
    def simpleName[A: Type]: String = {
      val colored = Type.prettyPrint[A]
      val mono = "\u001b\\[([0-9]+)m".r.replaceAllIn(colored, "")
      val start = mono.lastIndexOf(".") + 1
      val end = mono.indexOf("[", start) - 1
      mono.substring(start.max(0), if (end < 0) mono.length else end)
    }
```

which isn't optimal, but for a while it might be okayish.

**Reply by urmaul** [reply]
Oh great, thank you!

**Reply by MateuszKubuszok** [reply]
I added [some docs](https://chimney.readthedocs.io/en/latest/cookbook/#chimney-engine), and merged https://github.com/scalalandio/chimney/pull/574 - `chimney-engine` should get published on [Sonatype snapshots](https://github.com/scalalandio/chimney/actions/runs/11034274525/job/30647411003), but as I wrote in the added docs - this module should not be considered stable, nor its version a semantic version.

I don't expect many breaking changes, but some changes be necessary in the `TransformationContext` to unblock several features in the future. Maybe I'll find a way to not make them breaking, maybe not. 😅  It should not happen soon as I don't have much capacity for OSS lately, and I don't think recent merges would be published as a release, unless more people would contribute improvements in my stead.

**Reply by MateuszKubuszok** [reply]
`Derivation-engine` got released in [1.5.0](https://github.com/scalalandio/chimney/releases/tag/1.5.0) - as I wrote in the docs and changelog, the API might change (I think I might need to sometimes change something in the `TransformationContext` to implement the remaining feature requests from the backlog) but it shouldn't happen often.

**Reply by urmaul** [reply]
FYI the infinite implicit call loop is fixed now. I've added `ctx.config.isImplicitSummoningPreventedFor[From, To]` check in our custom implicit summoner rule.

**Comment by urmaul** [comment]
FYI I dropped the attempts when it turned out there's no easy way to handle forward- and backward-compatibility annotations the way teleproto did it. Also it turned out that using chimney directly for our stuff is much easier than a year ago so migrating our apps to use chimney is eatier than migrating teleproto.

Maybe I'll extract the prefixed proto enum migration but not sure if it's worth it.

Thank you for the help.
